### DISCUSSION TITLE
How to implement atom that must be initialized before consumed ?

### DISCUSSION THREAD

**Discussion by stevebeauge** [pmndrs/jotai#1897] [Q&A] [answered]
In my application, I have a complex component tree. I want to read an atom in part of the component where I am sure a value is set. If not, the app should crash because it's a bug.

A common use case that can illustrate this use case is the **logged user**. Imagine an atom containing the user name. Most of the application can be accessed only when the user is logged in. Component in this application part can safely expect a username is set.

In order to avoid having to check if the value is present every time I need the user name, I use this "pattern":

````typescript
const privateValueAtom = atom<string | undefined>(undefined);

const userNameAtom = atom(
  (get) => {
    const value = get(privateValueAtom);
    if (!value) throw new Error("Atom has not been initialized"); // if it happen, we're facing a highly critical bug
    return value;
  },
  (_, set, value: string) => {
    set(privateValueAtom, value);
  }
);
````

This way, I can consume the username safely:

````
const userName = useAtomValue(userNameAtom); // is string for sure
````

This works well but adds a bunch of boilerplate code, because every atom like this one needs private one, and a check to ensure the value is not undefined.

That's why I'm trying to implement some helper. My current attempt is:

````typescript
function atomWithRequiredCheck<Value>(initialValue?: Value | undefined) {
  const privateAtom = atom<Value | undefined>(initialValue);

  return atom(
    (get) => {
      const value = get(privateAtom);
      if (!value) throw new Error("Atom has not been initialized");
      return value;
    },
    (_, set, value: Value | undefined) => {
      set(privateAtom, value);
    }
  );
}
````

This helper is supposed to declare atom like this :
````typescript
const userNameChecked = atomWithRequiredCheck<string>();
const Uppercase: React.FC = () => {
  // Message is expected to always contains a value
  const message = useAtomValue(userNameChecked);
  return (
    <div>
      <p>{message.toUpperCase()}</p>
    </div>
  );
};
````

Either the atom contains a value or the components fails completely.

A full sample of what I'm trying to do is available here: https://codesandbox.io/s/hungry-shape-ydpnfl?file=/src/App.tsx:917-1952

My question is : am I doing things correctly? Is there better alternative? I also thought about creating a new hook `useCheckedAtomValue` that wrap the useAtomValue, but there are a bunch of existing hooks that should be wrapped.

My second question is : can I create atom within functions ? I think I can, because the function is called only once (in opposition to creating atoms within render function of react components)

**Comment by dai-shi** [comment, marked as answer]
> That's why I'm trying to implement some helper. My current attempt is:

That looks pretty good. I like it.

> My second question is : can I create atom within functions ?

Yes, you can. As you said, it works because the function is called just once (or limited times).
You might want read some code in `jotai/utils` and `jotaijs/*` packages and learn the pattern. (It doesn't mean they are everything. We may explore new patterns.)

**Reply by stevebeauge** [reply]
thx for your feedback. I'll continue in this direction.

**Comment by adrice727** [comment]
I'm glad I saw this because I'm trying to accomplish something similar.  @stevebeauge How are you handling errors in the case of an atom that hasn't been initialized?

**Reply by stevebeauge** [reply]
@adrice727 the check is in the get function of the nested atom within the atomWithRequiredCheck function
